src/prng/hc128.rs

```diff
@@ -0,0 +1,543 @@
+///
+/// [4]: Shashwat Raizada (January 2015),
+///      ["Some Results On Analysis And Implementation Of HC-128 Stream Cipher"]
+///      (http://library.isical.ac.in:8080/jspui/bitstream/123456789/6636/1/TH431.pdf).
+pub struct Hc128Rng {
+    state: Hc128,
+    results: [u32; 16],
+    index: usize,
+}
+
+impl Clone for Hc128Rng {
+    fn clone(&self) -> Hc128Rng {
+        Hc128Rng {
+            state: self.state,
+            results: self.results,
+            index: self.index
+        }
+    }
+}
+
+#[derive(Copy, Clone)]
+struct Hc128 {
+    t: [u32; 1024],
+    counter1024: usize,
+}
+
+// Custom Debug implementation that does not expose the internal state
+impl fmt::Debug for Hc128Rng {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Hc128Rng {{}}")
+    }
+}
+
+impl Hc128Rng {
+    pub fn init(seed: &[u32]) -> Hc128Rng {
+        #[inline]
+        fn f1(x: u32) -> u32 {
+            x.rotate_right(7) ^ x.rotate_right(18) ^ (x >> 3)
+        }
+
+        #[inline]
+        fn f2(x: u32) -> u32 {
+            x.rotate_right(17) ^ x.rotate_right(19) ^ (x >> 10)
+        }
+
+        let mut t = [0u32; 1024];
+
+        // Expand the key and iv into P and Q
+        let (key, iv) = seed.split_at(4);
+        t[..4].copy_from_slice(key);
+        t[4..8].copy_from_slice(key);
+        t[8..12].copy_from_slice(iv);
+        t[12..16].copy_from_slice(iv);
+
+        // Generate the 256 intermediate values W[16] ... W[256+16-1], and
+        // copy the last 16 generated values to the start op P.
+        for i in 16..256+16 {
+            t[i] = f2(t[i-2]).wrapping_add(t[i-7]).wrapping_add(f1(t[i-15]))
+                   .wrapping_add(t[i-16]).wrapping_add(i as u32);
+        }
+        {
+            let (p1, p2) = t.split_at_mut(256);
+            p1[0..16].copy_from_slice(&p2[0..16]);
+        }
+
+        // Generate both the P and Q tables
+        for i in 16..1024 {
+            t[i] = f2(t[i-2]).wrapping_add(t[i-7]).wrapping_add(f1(t[i-15]))
+                   .wrapping_add(t[i-16]).wrapping_add(256 + i as u32);
+        }
+
+        let mut state = Hc128Rng {
+            state: Hc128 { t: t, counter1024: 0 },
+            results: [0; 16],
+            index: 0,
+        };
+
+        // run the cipher 1024 steps
+        for _ in 0..64 { state.state.sixteen_steps() };
+        state.state.counter1024 = 0;
+
+        // Prepare the first set of results
+        state.state.update(&mut state.results);
+        state
+    }
+}
+
+impl Hc128 {
+    // One step of HC-128, update P and generate 32 bits keystream
+    #[inline(always)]
+    fn step_p(&mut self, i: usize, i511: usize, i3: usize, i10: usize, i12: usize)
+         -> u32
+    {
+        let (p, q) = self.t.split_at_mut(512);
+        // FIXME: it would be great if we the bounds checks here could be
+        // optimized out, and we would not need unsafe.
+        // This improves performance by about 7%.
```

Comment:
I have tried something like 20 variants. In some of them `p` and `q` where arrays, and `t` did not exist. If have tried macro's instead of functions, and various manual bounds checks in `update`. But no luck convincing LLVM they are unnecessary yet 😞 . I can't be much more precise, because I did not save or document all tries. Maybe expecting it to optimize out 7 bounds checks is to much to ask. It already reduces that down to 1 on average (but there is not any specific one, that is the average over the whole `update` function).